### ISSUE TITLE
attn_bias not populated in case of boolean mask in scaled_dot_product_attention

### ISSUE THREAD

**Issue by dhruvbird** [pytorch/pytorch#109072]
### 🐛 Describe the bug

Seems like the variable `attn_bias` is never populated in case of a `bool` attention mask.

https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py#L4923C1-L4927C39

```
if attn_mask is not None:
    if attn_mask.dtype == torch.bool:
        attn_mask.masked_fill_(attn_mask.logical_not(), float("-inf"))
    else:
        attn_bias += attn_mask
```

Maybe I'm missing something here.

I'm guessing that the code is mimicking the docs, so posted here. If the C++ implementation is different, please could you point me to it?

### Versions

master branch.

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by mikaylagawarecki** [comment]
If the mask is a boolean mask, my understanding is that the corresponding elements of the attention (`q @ k`) will be masked out before the `softmax`, since the `attn_mask` has `0` and `-inf` this is equivalent to adding the `attn_mask` to the output of `q @ k`

 I think the docs have a slight bug and should say

 ```python
 if attn_mask is not None:
    if attn_mask.dtype == torch.bool:
        attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
    else:
        attn_bias += attn_mask
```
